Add shell.nix

SimonAlling opened the pull request:
It's convenient for Nix(OS) users, who only have to run

    nix-shell

to get a sandbox with all dependencies available.

cxong commented:
I'm not familiar with NixOS; is this the best way to support it? Do the instructions described in https://github.com/opengaming/osgameclones#contributing not work? They are designed to be platform-agnostic so as long as you have python 2 + pipenv installed, the rest should be taken care for you. Does NixOS have another way to easily support pipenv projects, and not just osgc?

SimonAlling commented:
> I'm not familiar with NixOS; is this the best way to support it? Do the instructions described in https://github.com/opengaming/osgameclones#contributing not work? They are designed to be platform-agnostic so as long as you have python 2 + pipenv installed, the rest should be taken care for you. Does NixOS have another way to easily support pipenv projects, and not just osgc?

I can't say for sure that this is _the_ best way, but it's at least _a_ good way, often used on NixOS. By the way, it doesn't apply only to NixOS users, but to all users of the Nix package manager. In general, I believe Nix users are not very fond of installing dependencies (like Python 2) globally, but prefer to have them only where they're needed — e.g. in a `nix-shell`. (On NixOS, one installs packages globally by adding them to the declarative file `/etc/nixos/configuration.nix`, which describes the entire system configuration.)

As for the instructions in the readme, I can't get them to work. With Python 2 installed globally, I have run into two different error messages:

<details>
  <summary>In my working directory</summary>

```
$ pipenv install
Installing dependencies from Pipfile.lock (e72250)...
Traceback (most recent call last):▉▉▉▉▉ 0/11 — 00:00:00
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/bin/.pipenv-wrapped", line 12, in <module>
    sys.exit(cli())
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/cli.py", line 435, in install
    selective_upgrade=selective_upgrade,
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/core.py", line 1943, in do_install
    pypi_mirror=pypi_mirror,
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/core.py", line 1322, in do_init
    pypi_mirror=pypi_mirror,
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/core.py", line 807, in do_install_dependencies
    pypi_mirror=pypi_mirror,
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/core.py", line 1375, in pip_install
    package_name.split('--hash')[0].split('--trusted-host')[0]
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/requirementslib/models/requirements.py", line 704, in from_line
    line, extras = _strip_extras(line)
TypeError: 'module' object is not callable
  🐍   ▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉ 0/11 — 
```
</details>

<details>
  <summary>In a clean, freshly cloned repo</summary>

```
$ pipenv install
Creating a virtualenv for this project...
Pipfile: /home/alling/temp/osgameclones/Pipfile
Using /etc/profiles/per-user/alling/bin/python2.7 (2.7.15) to create virtualenv...
⠋/nix/store/80qva2zal82njkg4zvb7xg94mnikqksd-python3-3.6.8/bin/python3.6m: Error while finding module specification for 'pipenv.pew' (ModuleNotFoundError: No module named 'pipenv')

Virtualenv location: 
Traceback (most recent call last):
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/bin/.pipenv-wrapped", line 12, in <module>
    sys.exit(cli())
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/cli.py", line 435, in install
    selective_upgrade=selective_upgrade,
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/core.py", line 1759, in do_install
    pypi_mirror=pypi_mirror,
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/core.py", line 614, in ensure_project
    path_to_python = which('python') or which('py')
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/core.py", line 115, in which
    p = os.path.join(location, 'bin', command)
  File "/nix/store/80qva2zal82njkg4zvb7xg94mnikqksd-python3-3.6.8/lib/python3.6/posixpath.py", line 80, in join
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```
</details>

As you can see, Python 3 is mentioned in the error messages, probably because I (also) have Python 3 installed globally on my system. This is one of the main reasons why we prefer starting a `nix-shell` with local dependencies; then one can be sure that no globally installed package will interfere in any way.

All in all, the only way I've been able to build OSGC is with this `shell.nix` file. The `shellHook` part is not mine; I got it from another NixOS user when I asked for help with building OSGC. As for other ways to easily support pipenv projects, I don't know.

cxong commented:
It seems weird that pipenv would use python 3 here. I thought the `python_version` in pipfile (https://github.com/opengaming/osgameclones/blob/master/Pipfile#L16) would make it use python 2.

Does it work if you explicitly use pipenv with `--two`?

SimonAlling commented:
> Does it work if you explicitly use pipenv with `--two`?

<details>
    <summary>CLI example</summary>

```
$ python --version
Python 2.7.15
$ python2 --version
Python 2.7.15
$ python3 --version
Python 3.6.8
$ git clone git@github.com:opengaming/osgameclones
$ cd osgameclones
$ pipenv --two install
Creating a virtualenv for this project...
Pipfile: /home/alling/dev/osgameclones-new/Pipfile
Using /etc/profiles/per-user/alling/bin/python2 (2.7.15) to create virtualenv...
⠋/nix/store/80qva2zal82njkg4zvb7xg94mnikqksd-python3-3.6.8/bin/python3.6m: Error while finding module specification for 'pipenv.pew' (ModuleNotFoundError: No module named 'pipenv')

Virtualenv location: 
Traceback (most recent call last):
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/bin/.pipenv-wrapped", line 12, in <module>
    sys.exit(cli())
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 1063, in invoke
    Command.invoke(self, ctx)
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/vendor/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/cli.py", line 294, in cli
    three=three, python=python, warn=True, site_packages=site_packages, pypi_mirror=pypi_mirror
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/core.py", line 614, in ensure_project
    path_to_python = which('python') or which('py')
  File "/nix/store/jfpr7ndkncy7jb1jzxmqaics1grgcsxw-pipenv-2018.7.1/lib/python3.6/site-packages/pipenv/core.py", line 115, in which
    p = os.path.join(location, 'bin', command)
  File "/nix/store/80qva2zal82njkg4zvb7xg94mnikqksd-python3-3.6.8/lib/python3.6/posixpath.py", line 80, in join
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```
</details>

Note that Python 3.6 is still mentioned in the error message.

I don't know much about Python, its different versions and why these errors can occur if they are installed simultaneously. I know only one relevant thing: With `shell.nix`, it _just works™_ for Nix(OS) users.

SimonAlling commented:
What do you think, @cxong? This change would make it quite a lot easier for Nix users to contribute to OSGC. 🙂

cxong commented:
sorry, I'd rather not have platform-specific files in the project. We should move to python 3 soon so hopefully we won't need this.

cxong commented:
Please see #927 for a python3 migration
